src/theory/arith/nl/iand_solver.cpp

```diff
@@ -259,7 +246,53 @@ Node IAndSolver::sumBasedLemma(Node i)
   uint64_t granularity = options::BVAndIntegerGranularity();
   NodeManager* nm = NodeManager::currentNM();
   Node lem = nm->mkNode(
-      EQUAL, i, d_iandTable.createBitwiseNode(x, y, bvsize, granularity));
+      EQUAL, i, d_iandUtils.createSumNode(x, y, bvsize, granularity));
+  return lem;
+}
+
+Node IAndSolver::bitwiseLemma(Node i)
+{
+  Assert(i.getKind() == IAND);
+  Node x = i[0];
+  Node y = i[1];
+
+  unsigned bvsize = i.getOperator().getConst<IntAnd>().d_size;
+  uint64_t granularity = options::BVAndIntegerGranularity();
+
+  Rational absI = d_model.computeAbstractModelValue(i).getConst<Rational>();
+  Rational concI = d_model.computeConcreteModelValue(i).getConst<Rational>();
+
+  Assert(absI.isIntegral());
+  Assert(concI.isIntegral());
+
+  BitVector bvAbsI = BitVector(bvsize, absI.getNumerator());
+  BitVector bvConcI = BitVector(bvsize, concI.getNumerator());
+
+  NodeManager* nm = NodeManager::currentNM();
+  Node lem = d_true;
+
+  // compare each bit to bvI
+  Node cond;
+  Node bitIAnd;
+  unsigned high_bit;
+  for (unsigned j = 0; j < bvsize; j += granularity)
+  {
+    high_bit = j + granularity - 1;
+    // don't let high_bit pass bvsize
```

Comment:
Is it always enough to decrement by 1? In the sum-based scheme I have a decrement loop.
If it is enough, can you add an assertion that says that the difference is at most 1?

Reply:
Hmm... I'm not sure I follow. I'm probably missing something, what are you worried about? All I'm trying to handle is if your granularity surpasses the size when you count up. Say the granularity is 4 and this is bitwidth 6. Then first you get `extract 3 0`. Then you would get `extract 7 4` but then we just edit this to be `extract 5 4`.

Reply:
OK, I see. It should be OK. To make sure, maybe you could include tests with:
1. granularity 4 and bitwidth 6
2. granularity 5 and bitwidth 3